packages/components/login-widget/src/reducer.ts

```diff
@@ -0,0 +1,183 @@
+
+type Action = {
+  type: Event
+  payload?: any
+}
+
+function toWallet(address: string, chain: string, provider: any[]): Wallet {
+  const isInstalled = getProviders().isExtensionForChainInstalled(
+    chain.slice(0, 3).toUpperCase() as any,
+  )
+
+  const w: Wallet = {
+    address,
+    connectionStatus: 'not_installed',
+    providers: [...provider],
+    chainId: mainnetByChain[chain as SupportedChain],
+  }
+
+  if (isInstalled) {
+    w.connectionStatus = provider.length
+      ? 'connected_different_chain'
+      : 'disconnected'
+  }
+
+  if (
+    provider.length &&
+    isMainnetChecker[chain as SupportedChain](provider[0])
+  ) {
+    w.connectionStatus = 'connected'
+  }
+
+  return w
+}
+
+// preload chainId (because it is async, cannot place inside reducer)
+// before running the real action in reducer
+export const asyncActions = {
+  connect:
+    ({ dispatch }: any) =>
+    async (action: any) => {
+      const { provider } = action.payload
+      try {
+        const chainId = await provider.request({ method: 'eth_chainId' })
+        if (chainId) {
+          provider.chainId = chainId
+        }
+      } catch (e: any) {
+        console.error(e)
+      } finally {
+        dispatch({ type: 'do_connect', payload: action.payload })
+      }
+    },
+  login:
+    ({ dispatch }: any) =>
+    async (action: any) => {
+      const { provider } = action.payload
+      try {
+        const chainId = await provider.request({ method: 'eth_chainId' })
+        if (chainId) {
+          provider.chainId = chainId
+        }
+      } catch (e: any) {
+        console.error(e)
+      } finally {
+        dispatch({ type: 'do_login', payload: action.payload })
+      }
+    },
+}
+
+export type Reducer = (
+  state: LoginWidgetState,
+  action: Action,
+) => LoginWidgetState
+
+export function reducer(
+  state: LoginWidgetState,
+  action: Action,
+): LoginWidgetState {
+  switch (action.type) {
+    case 'do_connect': {
+      if (!state.isLoggedIn) return state
+
+      const { addresses, chain, provider } = action.payload
+      const newAddresses: string[] = []
+
+      // handle existing address first
+      for (const addr of addresses) {
+        const w = state.wallets.find(
+          (w) => w.address.toLowerCase() === addr.toLowerCase(),
+        )
+        if (!w) {
+          newAddresses.push(addr)
+          continue
+        }
+
+        const updated = toWallet(w.address, chain, [provider])
+        w.providers = updated.providers
+        w.connectionStatus = updated.connectionStatus
+      }
+
+      // now handle new addresses
+      const newState = state.wallets.concat(
+        newAddresses.map((na) => toWallet(na, chain, [provider])),
+      )
+
+      return {
+        ...state,
+        wallets: newState,
+      }
+    }
+
+    case 'logout': {
+      return {
+        ...state,
+        isLoggedIn: false,
+        wallets: [],
+      }
+    }
+
+    case 'do_login': {
```

Comment:
// do_login initially fetch a list of user associated wallets, then map a logged in wallet as 'connected'